Collection Sheet Enhancement

Aditya3815 opened the pull request:
https://github.com/user-attachments/assets/29b32ce2-4805-4a7a-a411-9265226afe57

Fixes - [Jira-#MIFOSAC-453](https://mifosforge.jira.com/browse/MIFOSAC-453)

Didn't create a Jira ticket, click [here](https://mifosforge.jira.com/jira/software/c/projects/MIFOSAC/issues/) to create new.

Please Add Screenshots If there are any UI changes.

| Before                                     | After                                  |
|--------------------------------------------|----------------------------------------|
|                                            |                                        |

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.

Aditya3815 committed:
start collection-sheet enhancement

Aditya3815 committed:
Add Preview dependencies and annotation DevicePreview to Preview

Aditya3815 committed:
Index 1 out of bounds for length 1 resolved

itsPronay commented:
what is the purpose of this PR? Can you mention in the ticket?

Aditya3815 committed:
Merge branch 'refs/heads/kmp-impl' into collectionsheet-enhancement

# Conflicts:
#	cmp-android/dependencies/demoDebugRuntimeClasspath.txt
#	cmp-android/dependencies/prodDebugRuntimeClasspath.txt

Aditya3815 committed:
updated

Aditya3815 committed:
Merge branch 'refs/heads/kmp-impl' into collectionsheet-enhancement

# Conflicts:
#	feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/paymentDetails/PaymentDetailsScreen.kt

Aditya3815 committed:
Complete

revanthkumarJ commented:
@Aditya3815 can you replace if Total Charges , and other fields if null show zero on screen as showing null on screen to user is not appropriate

Aditya3815 committed:
complete 2

HekmatullahAmin commented on feature/collectionSheet/build.gradle.kts:
```diff
@@ -25,11 +26,16 @@ kotlin {
             implementation(compose.ui)
             implementation(projects.core.domain)
             implementation(libs.kotlinx.serialization.json)
+            implementation(compose.components.uiToolingPreview)
```
Remove this as u already defined in `dependencies {}` block

HekmatullahAmin replied:
i mentioned remove it from `commonMain` but keep the `dependencies{}`:
- cause when we create a project from `KMP wizard` this way it is done by default maybe they have sth in mind where in future the previews will work for all platform if we have it here
```
dependencies {
    debugImplementation(compose.uiTooling)
}
```

Aditya3815 committed:
Collection Sheet Enhancement Complete

HekmatullahAmin commented on feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/individualCollectionSheetDetails/IndividualCollectionSheetDetailsScreen.kt:
```diff
@@ -210,6 +210,8 @@ private fun IndividualCollectionSheetItem(
     modifier: Modifier = Modifier,
     onClick: () -> Unit,
 ) {
+    val loan = client.loans?.getOrNull(index)
```
i see that this index is not use anywhere except finding the loan. so instead of passing index why not passing the loan itself from place this composable is called

HekmatullahAmin commented on feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/newIndividualCollectionSheet/NewIndividualCollectionSheetScreen.kt:
```diff
@@ -229,12 +233,24 @@ internal fun NewIndividualCollectionSheetScreen(
-                        state.officeList[index].id.let {
-                            getStaffList(it)
-                            officeId = it
+                        println("DEBUG: Office selection - Index: $index, List size: ${state.officeList.size}, Value: $value")
```
remove all of these `println`

HekmatullahAmin commented on feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/individualCollectionSheetDetails/IndividualCollectionSheetDetailsScreen.kt:
```diff
@@ -246,18 +248,16 @@ private fun IndividualCollectionSheetItem(
                     Spacer(modifier = Modifier.width(16.dp))
                     Text(
-                        text = client.loans?.get(index)?.totalDue.toString(),
+                        text = (loan?.totalDue ?: 0).toString(),
```
- the zero should be in this format `0.0`
- i see this screen there are things that need to be taken care of at viewmodel and the screen should only display to user, but no worries for now we should come back for all of these

HekmatullahAmin commented on feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/newIndividualCollectionSheet/NewIndividualCollectionSheetScreen.kt:
```diff
@@ -254,10 +270,21 @@ internal fun NewIndividualCollectionSheetScreen(
-                        state.staffList[index].id?.let {
-                            staffId = it
+                        try {
+                            if (index >= 0 && index < state.staffList.size) {
```
having these codes like `try/cathc` is highly discouraged in UI. use build it extension functions: sth like below change accordingly
```
state.officeList.getOrNull(index)?.let { selectedOfficeEntity ->
    selectedOfficeEntity.id?.let {
        getStaffList(it)
        officeId = it
    }
    selectedOffice = selectedOfficeEntity.name.toString()
}
```

Aditya3815 replied:
@HekmatullahAmin as you requested I have done changes can you do review now

Aditya3815 replied:
After that I will run the spotlessApply

Aditya3815 committed:
Index removed as parameter instead directly passing loan
Removed duplicate depndencies from build.gradle.kts
change the Int to Double type 0 -> 0.0

HekmatullahAmin commented on feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/individualCollectionSheetDetails/IndividualCollectionSheetDetailsScreen.kt:
```diff
@@ -178,17 +180,17 @@ internal fun IndividualCollectionSheetDetailsScreen(
                                     sheet.paymentTypeOptions?.let { paymentTypeOptions ->
                                         submit(
-                                            index,
+                                            0,
```
- By removing index i meant to remove it from composable and just pass the loan as a parameter instead of going to composable and finding again using the index. where the index is not used in that composable at any line.
- but you removed the index from submit and other areas an hardcoded it. 
- when i review the code i compare it to the previous code cause i am not fully aware of each variable and what specefic screen do. so keep the things as before just inside the `IndividualCollectionSheetItem` instead of passing index pass the loan using maybe `client.loans?.get(index)`.
- so have `itemsIndexed` as before

HekmatullahAmin commented on feature/collectionSheet/src/commonMain/kotlin/com/mifos/feature/individualCollectionSheet/individualCollectionSheetDetails/IndividualCollectionSheetDetailsScreen.kt:
```diff
@@ -178,17 +180,17 @@ internal fun IndividualCollectionSheetDetailsScreen(
             } else {
                 LazyColumn {
```
whenever u use `LazyColumn` use `key` for better performance

revanthkumarJ committed:
Merge branch 'kmp-impl' into collectionsheet-enhancement

revanthkumarJ committed:
Merge branch 'kmp-impl' into collectionsheet-enhancement

revanthkumarJ committed:
updated

revanthkumarJ committed:
updated

revanthkumarJ committed:
updated

revanthkumarJ committed:
updated

revanthkumarJ committed:
updated

niyajali committed:
Merge branch 'kmp-impl' into collectionsheet-enhancement

niyajali committed:
Merge branch 'kmp-impl' into collectionsheet-enhancement

Nagarjuna0033 commented:
@Aditya3815 are you working on this pr, if not could you close this
